### PR TITLE
Add a printer that knows how to print user-defined columns

### PR DESCRIPTION
**Pull request by brendandburns**
@bgrant0607 @jlowdermilk @JanetKuo @deads2k 

This was discussed in the context of https://github.com/kubernetes/kubernetes/issues/7843

I can't find a dedicated issue for it, but I swear we discussed it.

I'll integrate this as a supported printer in a follow-on PR.

### CODE REVIEW

File: pkg/kubectl/custom_column_printer.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+// CustomColumnPrinter is a printer that knows how to print arbitrary columns
+// of data from templates specified in the `Columns` array
+type CustomColumnsPrinter struct {
+	Columns []Column
+}
+
+func (s *CustomColumnsPrinter) PrintObj(obj runtime.Object, out io.Writer) error {
+	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
+	buff := &bytes.Buffer{}
```
Comment (deads2k) [inline pkg/kubectl/custom_column_printer.go]: I don't see this used.

File: pkg/kubectl/custom_column_printer.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	for ix := range s.Columns {
+		fmt.Fprintf(w, "%s\t", s.Columns[ix].Header)
+	}
+	fmt.Fprintln(w, buff.String())
```
Comment (deads2k) [inline pkg/kubectl/custom_column_printer.go]: This is just `fmt.Fprintln(w)`, right?

File: pkg/kubectl/custom_column_printer.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+func (s *CustomColumnsPrinter) printOneObject(obj runtime.Object, parsers []*jsonpath.JSONPath, out io.Writer) error {
+	for ix := range parsers {
+		parser := parsers[ix]
+		values, err := parser.FindResults(reflect.ValueOf(obj).Elem().Interface())
+		if err != nil {
+			return err
+		}
+		if len(values) != 1 || len(values[0]) != 1 {
+			return fmt.Errorf("unexpected number of values: %d %v", len(values), values)
```
Comment (deads2k) [inline pkg/kubectl/custom_column_printer.go]: I'm not certain that I always want this condition to fail hard.  If I want to list out the first, second, and third items in a list (say the secrets on a serviceaccount), not all serviceaccounts will have at least three, but I probably don't want my printer to fail to display anything to me.

**Comment by bgrant0607** [comment]
See also #11695

cc @kubernetes/kubectl @mwielgus

**Comment by mwielgus** [comment]
Can we combine HumanReadablePrinter (from #11695) and CustomColumnsPrinter into one?

**Comment by brendandburns** [comment]
Yeah, I think that's the right thing to do in the limit, have a bunch of template files as @bgrant0607 suggests and refactor HumanReadablePrinter in terms of the CustomColumnsPrinter.

Why don't we merge in the custom columns printer, and then you can refactor #11695 in terms of that printer?  Does that work for you?

File: pkg/kubectl/custom_column_printer.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/custom_column_printer.go]: oops, first implementation cruft, removed.

File: pkg/kubectl/custom_column_printer.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/custom_column_printer.go]: as above, fixed.

File: pkg/kubectl/custom_column_printer.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/custom_column_printer.go]: Much improved. thanks for the comment.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by deads2k** [comment]
lgtm.  Weird that `test-cmd` would only fail on go 1.3.  I don't have re-run rights, can you kick shippable?

File: pkg/kubectl/custom_column_printer_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	tests := []struct {
+		columns        []Column
+		obj            runtime.Object
+		expectedOutput string
+	}{
+		{
+			columns: []Column{
+				{
+					Header:    "NAME",
+					FieldSpec: "{.ObjectMeta.Name}",
```
Comment (bgrant0607) [inline pkg/kubectl/custom_column_printer_test.go]: These expressions should use the json field names, not the Go field names.

Comment (bgrant0607) [inline pkg/kubectl/custom_column_printer_test.go]: They also need to reference the versioned APIs.

Comment (deads2k) [inline pkg/kubectl/custom_column_printer_test.go]: > They also need to reference the versioned APIs.

That being the case, the `SortingPrinter` from https://github.com/kubernetes/kubernetes/pull/12432 needs updating as well.

**Comment by bgrant0607** [comment]
This implementation uses the jsonpath template syntax, which is more usable for raw fields than the Go template syntax, but is insufficient for the HumanReadablePrinter use case.

Note: jsonpath template support is being added to kubectl in PR #12202.

**Comment by bgrant0607** [comment]
Any thoughts about how this would be exposed through kubectl? Assuming we continue to support both Go templates and jsonpath templates, we'll want this functionality for both template schemes.

**Comment by brendandburns** [comment]
The jsonPath should use json names, but that requires changes to the JSONPath libraries reflection code.

Why is this insufficient for human readable printer?  Just for human readable numbers like age, etc.

File: pkg/kubectl/custom_column_printer_test.go (hunk 4 above)
Comment (brendandburns) [inline pkg/kubectl/custom_column_printer_test.go]: How do you imaging referencing versioned apis should work?  I don't think that's right.  I think that the templates themselves may be versioned, and the user will select, but in this context, I don't think that versioning has a place.

**Comment by brendandburns** [comment]
For kubectl, I think that we want to both have:
`kubectl ... --custom-columns=<template>:<header>,<template>:<header>,<template>:<header>`

as well as:

`kubectl ... --custom-columns-template=<template-file>`

File: pkg/kubectl/custom_column_printer_test.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/kubectl/custom_column_printer_test.go]: Go field names are an implementation detail that we shouldn't expose to users. The json field names are the API, and all such API resources and their fields are versioned. We already do this with the Go templates.

**Comment by bgrant0607** [comment]
TestPrintJSONPath in #12202 prints `"{.metadata.name}"`, which refers to the json fields of a versioned API resource.

**Comment by bgrant0607** [comment]
Both from command line and file make sense, but should be output format options, not new flags.

File: pkg/kubectl/custom_column_printer_test.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/kubectl/custom_column_printer_test.go]: Good call, @deads2k. Yes, the sorting feature does need to be fixed.

**Comment by bgrant0607** [comment]
As for HumanReadablePrinter, many of the fields (e.g., time fields) are transformed for output. How would such transformations be specified when using jsonpath templates?

**Comment by brendandburns** [comment]
For jsonpath templates, we could either have users specify explicit output transformations (kind of ackward) or infer them from the type of the value (far easier to use)

Either way, it's the right approach from technical organization.

Ah, I think I misinterpretted your comments, I think you meant: "You should run templates against versioned objects"  not "the templates should include the API version"

I think that's probably true, but it is a different from other printers, we'll have to think about the right way to expose this.

**Comment by brendandburns** [comment]
Ok, this is fixed to use json names, and I fixed the sorting printer too (and I fixed up some hacky code in the jsonPath printer that I should have caught before...)

Please take another look.

Thanks!
--brendan

File: pkg/kubectl/custom_column_printer_test.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			expectedOutput: `NAME
+foo
+bar
+`,
+		},
+		{
+			columns: []Column{
+				{
+					Header:    "NAME",
+					FieldSpec: "{.ObjectMeta.Name}",
```
Comment (bgrant0607) [inline pkg/kubectl/custom_column_printer_test.go]: Why are these not the json field names?

**Comment by bgrant0607** [comment]
It shouldn't be different than -o template. That definitely uses the versioned json field names. We had to add --api-version= to all our scripts that used templates to prevent them from breaking repeatedly when changing the default API version.

File: pkg/kubectl/custom_column_printer_test.go (hunk 5 above)
Comment (brendandburns) [inline pkg/kubectl/custom_column_printer_test.go]: oops, oversight.  fixed.

**Comment by brendandburns** [comment]
@bgrant0607 yeah, I guess you're right, I always thought we took it into the in-memory representation and then printed them out, but after looking at the code, I realize I'm wrong.

Anyway, this works correctly now (as does the sorting printer)

and `jsonpath` should now correctly handle `inline` fields, which I'm not sure it did before...

**Comment by bgrant0607** [comment]
@jlowdermilk PTAL

File: pkg/kubectl/custom_column_printer.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+// CustomColumnPrinter is a printer that knows how to print arbitrary columns
+// of data from templates specified in the `Columns` array
+type CustomColumnsPrinter struct {
+	Columns []Column
+}
+
+func (s *CustomColumnsPrinter) PrintObj(obj runtime.Object, out io.Writer) error {
+	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
```
Comment (jlowdermilk) [inline pkg/kubectl/custom_column_printer.go]: Suggest making these params constants so it's easy to see what each one is.

File: pkg/util/jsonpath/jsonpath.go [hunk 7]
```diff
@@ -258,6 +259,38 @@ func (j *JSONPath) evalUnion(input []reflect.Value, node *UnionNode) ([]reflect.
...
+		if parts[0] == node.Value {
+			return value.Field(ix), nil
+		}
+		if len(parts[0]) == 0 {
+			val := value.Field(ix)
+			inlineValue = &val
+		}
+	}
+	if inlineValue != nil {
+		fmt.Printf("inline value while looking for: %s\n", node.Value)
```
Comment (jlowdermilk) [inline pkg/util/jsonpath/jsonpath.go]: This looks like a debug line.

File: pkg/kubectl/custom_column_printer.go (hunk 6 above)
Comment (kargakis) [inline pkg/kubectl/custom_column_printer.go]: +1

Comment (brendandburns) [inline pkg/kubectl/custom_column_printer.go]: done.

File: pkg/util/jsonpath/jsonpath.go (hunk 7 above)
Comment (brendandburns) [inline pkg/util/jsonpath/jsonpath.go]: removed.

**Comment by brendandburns** [comment]
Comments addressed, @jlowdermilk ptal

thanks
--brendan

**Comment by brendandburns** [comment]
nm, I'm fixing the test...

**Comment by jlowdermilk** [comment]
lgtm, please squash

**Comment by brendandburns** [comment]
squashed.

**Comment by brendandburns** [comment]
LGTM was before last commit, removing LGTM

**Comment by brendandburns** [comment]
last push was a rebase...  Re-adding LGTM

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by brendandburns** [comment]
@k8s-bot test this please

**Comment by jlowdermilk** [comment]
lgtm

File: pkg/kubectl/custom_column_printer.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+			if err := s.printOneObject(objs[ix], parsers, w); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := s.printOneObject(obj, parsers, w); err != nil {
+			return err
+		}
+	}
+	return w.Flush()
```
Comment (kargakis) [inline pkg/kubectl/custom_column_printer.go]: Shouldn't this flush be deferred?

**Comment by brendandburns** [comment]
merging, since submit-queue skipped this PR due to github API issues.

**Pull request merged**